programs/bpf_loader/src/lib.rs

```diff
@@ -371,13 +379,94 @@ pub fn create_vm<'a, 'b>(
             allocator.clone(),
         )
         .map_err(SyscallError::InstructionError)?;
-    let result = EbpfVm::new(
-        program,
-        invoke_context,
-        allocator.borrow_mut().get_heap(),
+    let stack_len = stack.len();
+    let accounts = Arc::clone(invoke_context.transaction_context.accounts());
+    let memory_mapping = create_memory_mapping(
+        program.get_executable(),
+        stack,
+        allocator.borrow_mut().heap_mut(),
         regions,
-    );
-    result
+        Some(Box::new(move |index_in_transaction| {
+            // The two calls below can't relly fail. If they fail because of a bug,
+            // whatever is writing will trigger an EbpfError::AccessViolation like
+            // if the region was readonly, and the transaction will fail gracefully.
+            let mut account = accounts
+                .try_borrow_mut(index_in_transaction as IndexOfAccount)
+                .map_err(|_| ())?;
+            accounts
+                .touch(index_in_transaction as IndexOfAccount)
+                .map_err(|_| ())?;
+
+            if account.is_shared() {
+                // See BorrowedAccount::make_data_mut() as to why we reserve extra
+                // MAX_PERMITTED_DATA_INCREASE bytes here.
+                account.reserve(MAX_PERMITTED_DATA_INCREASE);
+            }
+            Ok(account.data_as_mut_slice().as_mut_ptr() as u64)
+        })),
+    )?;
+
+    EbpfVm::new(program, invoke_context, memory_mapping, stack_len)
+}
+
+#[macro_export]
+macro_rules! create_vm {
```

Comment:
Why do we need a function and a macro?

Reply:
Among other things, rust-analyzer support for macro definitions is limited. fmt-on-save and
navigation for example don't work in this definition. Here the macro is needed only to create the
bindings, having the bulk of the code in an actual function makes it less annoying to work with.